src/panic.rs

```diff
@@ -16,15 +17,46 @@ Python interpreter to exit.
 );
 
 impl PanicException {
-    // Try to format the error in the same way panic does
+    /// Creates a new PanicException from a panic payload.
+    ///
+    /// Attempts to format the error in the same way panic does.
+    ///
+    /// If dropping `payload` panics, will fatally abort.
     #[cold]
     pub(crate) fn from_panic_payload(payload: Box<dyn Any + Send + 'static>) -> PyErr {
-        if let Some(string) = payload.downcast_ref::<String>() {
-            Self::new_err((string.clone(),))
-        } else if let Some(s) = payload.downcast_ref::<&str>() {
-            Self::new_err((s.to_string(),))
+        let err = if let Some(msg) = panic_payload_to_string(payload) {
+            Self::new_err((msg,))
         } else {
             Self::new_err(("panic from Rust code",))
-        }
+        };
+        err
     }
 }
+
+/// Formats a panic payload into a string, and drops the panic payload.
+///
+/// If dropping the panic payload panics, aborts the process.
+pub(crate) fn panic_payload_to_string(payload: Box<dyn Any + Send + 'static>) -> Option<String> {
+    let msg = if let Some(string) = payload.downcast_ref::<String>() {
+        Some(string.clone())
+    } else if let Some(s) = payload.downcast_ref::<&str>() {
+        Some(s.to_string())
+    } else {
+        None
+    };
+    abort_on_unwind(move || drop(payload), "panic while dropping panic payload");
+    msg
+}
+
+/// Executes `f()`. If it panics, attempts to write `msg` to stderr and then aborts.
+pub(crate) fn abort_on_unwind<F: FnOnce()>(f: F, msg: &str) {
+    // Always fine to use AssertUnwindSafe here because a panic will abort.
+    std::panic::catch_unwind(std::panic::AssertUnwindSafe(f))
+        .unwrap_or_else(|_| abort_with_msg(msg));
+}
+
+/// Attempts to write `msg` to stderr, then aborts.
+pub(crate) fn abort_with_msg(msg: &str) -> ! {
+    let _ = writeln!(std::io::stderr(), "FATAL: {}, aborting", msg);
```

Comment:
I think the `Stderr` wrapper returned by `stderr` used a `StderrLock` internally, the [`Write` impl of which calls `borrow_mut`](https://doc.rust-lang.org/stable/std/io/struct.StderrLock.html#impl-Write) so it seems this could panic itself. (The comment removed in `src/types/function.rs` seems to indicate this as well.)

Could this be a problem here since we are already unwinding? Is there a way to portably avoid calling into the whole `Stdio` locking machinery and the process go down my reliably but still make a best effort to print the message?

Is there

Reply:
Since we already depend on `libc`, would calling `write` with `STDERR_FILENO` be portable? Or is this just being paranoid and overdoing it?

Reply:
I looked into how std does this and they have a `rtabort` macro which calls into `rtprintpanic` which uses the normal `format!` machinery

```rust
macro_rules! rtprintpanic {
    ($($t:tt)*) => {
        if let Some(mut out) = crate::sys::stdio::panic_output() {
            let _ = crate::io::Write::write_fmt(&mut out, format_args!($($t)*));
        }
    }
}
```

but writes into a special `panic_output` which for example on Unix-like platforms is

```rust
impl Stderr {
    pub const fn new() -> Stderr {
        Stderr(())
    }
}

impl io::Write for Stderr {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        unsafe { ManuallyDrop::new(FileDesc::from_raw_fd(libc::STDERR_FILENO)).write(buf) }
    }
}
```

(It is a bit more complicated on Windows since it handles UTF-8 to UTF-16 conversion, but since our messages are always plain ASCII, I think the Unix approach should generally work for us.)

Reply:
Actually, I think we can let std handle all of this. My initial idea above was to use abort-due-to-panic-during-unwind, i.e. `PanicTrap::drop` panics instead of calling `abort_with_msg` which means std will take care of writing both panic messages and a backtrace without use having to muck around in the low-level details, c.f. [Rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b75409ec1b4117f27e7da49c2451d411).

Reply:
This sounds like a very sensible suggestion, I'll push an amendment shortly!